File: src/prod/src/Reliability/Replication/ReplicaInformation.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

namespace Reliability {
namespace ReplicationComponent {

using std::wstring;
using Common::Assert;

ReplicaInformation::ReplicaInformation(
    FABRIC_REPLICA_ID const & id, 
    ::FABRIC_REPLICA_ROLE const & role,
    wstring const & replicatorAddress,
    bool mustCatchup,
    FABRIC_SEQUENCE_NUMBER currentProgress,
    FABRIC_SEQUENCE_NUMBER catchupCapability)
    :   info_(), 
        ex1Info_(),
        replicatorAddress_(replicatorAddress),
        transportId_()
{
    info_.Id = id;
    info_.Role = role;
    info_.ReplicatorAddress = replicatorAddress_.c_str();
    info_.CurrentProgress = currentProgress;
    info_.CatchUpCapability = catchupCapability;
    info_.Reserved = &ex1Info_;
    ex1Info_.MustCatchup = mustCatchup;
    ex1Info_.Reserved = nullptr;
}

ReplicaInformation::ReplicaInformation(FABRIC_REPLICA_INFORMATION const * value)
    : info_(*value),
    ex1Info_(*(static_cast<FABRIC_REPLICA_INFORMATION_EX1*>(value->Reserved))),
    replicatorAddress_(),
    transportId_()
{
    ASSERT_IF(ex1Info_.Reserved != NULL, "FABRIC_REPLICA_INFORMATION_EX1 Reserved must be null.");

    ASSERT_IF(value->ReplicatorAddress == NULL, "ReplicatorAddress cannot be null.");
    wstring endpoint = wstring(value->ReplicatorAddress);
    wstring::size_type index =  endpoint.find(Constants::ReplicationEndpointDelimiter);
    ASSERT_IF(index == wstring::npos, "ReplicatorAddress {0} should contain delimiter {1}.", endpoint, Constants::ReplicationEndpointDelimiter);
    index =  endpoint.find(Constants::ReplicationEndpointDelimiter, index+1);
    ASSERT_IF(index == wstring::npos, "ReplicatorAddress {0} should contain delimiter {1} before incarnationId.", endpoint, Constants::ReplicationEndpointDelimiter);
    replicatorAddress_ = endpoint.substr(0, index);
    transportId_ = endpoint.substr(index + 1, endpoint.size() - (index + 1));
    info_.ReplicatorAddress = endpoint.c_str();
}

ReplicaInformation::ReplicaInformation(ReplicaInformation const & other)
    : info_(other.info_),
    ex1Info_(other.ex1Info_),
    replicatorAddress_(other.replicatorAddress_),
    transportId_(other.transportId_)
{
    info_.ReplicatorAddress = replicatorAddress_.c_str();
}

ReplicaInformation::ReplicaInformation(ReplicaInformation && other)
    : info_(other.info_),
    ex1Info_(other.ex1Info_),
    replicatorAddress_(std::move(other.replicatorAddress_)),
    transportId_(std::move(other.transportId_))
{
    info_.ReplicatorAddress = replicatorAddress_.c_str();
}

ReplicaInformation::~ReplicaInformation()
{
}

void ReplicaInformation::WriteTo(Common::TextWriter& w, Common::FormatOptions const&) const
{
    w.Write("{0}@{1}: {2}, {3} {4}", Id, ReplicatorAddress, Role, CurrentProgress, CatchUpCapability);
}

} // end namespace ReplicationComponent
} // end namespace Reliability

File: src/prod/src/managed/Api/src/System/Fabric/ReplicaInformation.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace System.Fabric
{
    using System;
    using System.Fabric.Common;
    using System.Fabric.Interop;

    /// <summary>
    /// This supports the Service Fabric infrastructure and is not meant to be used directly from your code.
    /// </summary>
    public sealed class ReplicaInformation
    {
        /// <summary>
        /// This supports the Service Fabric infrastructure and is not meant to be used directly from your code.
        /// </summary>
        public ReplicaInformation()
        {
        }

        /// <summary>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</summary>
        /// <value>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</value>
        public long Id
        {
            get;
            private set;
        }

        /// <summary>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</summary>
        /// <value>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</value>
        public ReplicaRole Role
        {
            get;
            private set;
        }

        /// <summary>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</summary>
        /// <value>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</value>
        public ReplicaStatus Status
        {
            get;
            private set;
        }

        /// <summary>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</summary>
        /// <value>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</value>
        public string ReplicatorAddress
        {
            get;
            private set;
        }

        /// <summary>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</summary>
        /// <value>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</value>
        public long CatchUpCapability
        {
            get;
            private set;
        }

        /// <summary>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</summary>
        /// <value>This supports the Service Fabric infrastructure and is not meant to be used directly from your code.</value>
        public long CurrentProgress
        {
            get;
            private set;
        }

        /// <summary>
        /// FOR INTERNAL USE ONLY.
        /// Gets a value indicating whether the replica must be caught up as part of a WaitForQuorumCatchup.
        /// </summary>
        /// <value>Whether the replica must be caught up as part of a WaitForQuorumCatchup.</value>
        public bool MustCatchup
        {
            get;
            private set;
        }

        internal static unsafe ReplicaInformation[] FromNativeArray(uint replicaCount, IntPtr replicaInfosIntPtr)
        {
            ReplicaInformation[] replicaInfos = new ReplicaInformation[(int)replicaCount];
            var nativeInfos = (NativeTypes.FABRIC_REPLICA_INFORMATION*)replicaInfosIntPtr;

            for (int i = 0; i < replicaInfos.Length; i++)
            {
                replicaInfos[i] = ReplicaInformation.FromNative((IntPtr)(&nativeInfos[i]));
            }

            return replicaInfos;
        }

        internal static unsafe ReplicaInformation FromNative(IntPtr replicaInfoIntPtr)
        {
            var nativeInfo = (NativeTypes.FABRIC_REPLICA_INFORMATION*)replicaInfoIntPtr;
            var info = new ReplicaInformation();

            info.Id = nativeInfo->Id;
            info.Role = (System.Fabric.ReplicaRole)nativeInfo->Role;
            info.Status = (System.Fabric.ReplicaStatus)nativeInfo->Status;
            info.ReplicatorAddress = NativeTypes.FromNativeString(nativeInfo->ReplicatorAddress);
            info.CatchUpCapability = nativeInfo->CatchUpCapability;
            info.CurrentProgress = nativeInfo->CurrentProgress;

            ReleaseAssert.AssertIf(nativeInfo->Reserved == IntPtr.Zero, "ReplicaInfo->Reserved cannot be null");

            var ex1 = (NativeTypes.FABRIC_REPLICA_INFORMATION_EX1*) nativeInfo->Reserved;
            info.MustCatchup = NativeTypes.FromBOOLEAN(ex1->MustCatchup);

            return info;
        }

        internal void ToNative(PinCollection pin, out NativeTypes.FABRIC_REPLICA_INFORMATION native)
        {
            native.Id = this.Id;
            native.Role = (NativeTypes.FABRIC_REPLICA_ROLE)this.Role;
            native.Status = (NativeTypes.FABRIC_REPLICA_STATUS)this.Status;
            native.ReplicatorAddress = pin.AddBlittable(this.ReplicatorAddress);
            native.CatchUpCapability = this.CatchUpCapability;
            native.CurrentProgress = this.CurrentProgress;


            var ex1 = new NativeTypes.FABRIC_REPLICA_INFORMATION_EX1[1];
            ex1[0].MustCatchup = NativeTypes.ToBOOLEAN(this.MustCatchup);
            ex1[0].Reserved = IntPtr.Zero;
            native.Reserved = pin.AddBlittable(ex1);
        }
    }
}